akka-http-server/src/main/scala/endpoints/akkahttp/routing/BasicAuthentication.scala

```diff
@@ -0,0 +1,36 @@
+package endpoints.akkahttp.routing
+
+import java.util.Base64
+
+import akka.http.scaladsl.model.headers.{Authorization, BasicHttpCredentials}
+import akka.http.scaladsl.model.{HttpHeader, HttpResponse, StatusCodes}
+import akka.http.scaladsl.server.Directives
+import endpoints.algebra
+import endpoints.algebra.BasicAuthentication.Credentials
+
+trait BasicAuthentication extends algebra.BasicAuthentication with Endpoints {
+
+  /**
+    * Extracts the credentials from the request headers.
+    * In case of absence of credentials rejects request
+    */
+  private[endpoints] lazy val basicAuthentication: RequestHeaders[Credentials] =
+    Directives.headerValue(extractCredentials)
```

Comment:
I think that defining in which conditions the routing process should try the next endpoint or return a failure is a matter of choice and several options can be valid, we just have to agree on a choice because clients will expect servers to behave consistently. I admit that I probably biased the `play.routing` semantics towards the current behavior of Play framework.
However, I still think that in the case of Basic authentication it makes sense to return a 401 status if the header is absent. Web browsers rely on this status and prompt the user for a login and a password. If you don’t return the 401 status then users can not supply their credentials if they use the API from their browser. You can try to run `+example-basic-server/run` and then open http://localhost:9000/auth in your browser to see what I mean.
So, my point is that we should not completely ignore web standards and always use 200 as a status code, for instance. We should take existing HTTP clients into account.